### PR TITLE
Environment pinning

### PR DESCRIPTION
**Pull request by guyer**
Use [`conda-lock`](https://github.com/conda/conda-lock) to pin environments.

This PR:
- tests every commit for breaking changes in the commit
- tests `master` nightly at midnight for breaking changes in dependencies

### CODE REVIEW

File: docs/source/ADMINISTRATA.rst [hunk 1]
```diff
@@ -119,6 +119,29 @@ We use the :term:`Azure` and :term:`GitHub Actions` cloud services for
...
 .. _.github/workflows/nix.yml: https://github.com/usnistgov/fipy/blob/master/.github/workflows/nix.yml
 
+===============
+Conda Lockfiles
+===============
+
+The `conda-lock <https://github.com/conda/conda-lock>`_ lockfiles in
+:file:`environments/locks/` can be updated with::
+
+    $ for solver in petsc pysparse scipy trilinos
```
Comment (tkphd) [inline docs/source/ADMINISTRATA.rst:129]: Refactor as a rst code-block.

File: INSTALLATION.rst [hunk 2]
```diff
@@ -61,59 +63,89 @@ Recommended Method
...
-   These days, the most useful channel is `conda-forge`, which provides
+   These days, the most useful channel is conda-forge_, which provides
    everything necessary to install :term:`FiPy`.
 
 * `install Miniconda`_ on your computer
-* run::
+* Install :term:`FiPy` :ref:`REQUIREDPACKAGES` with one of the following:
 
-    $ conda create --name <MYFIPYENV> --channel conda-forge python=<PYTHONVERSION> fipy gmsh
+    * Conda environment files::
```
Comment (wd15) [inline INSTALLATION.rst:72]: This option will pin all the packages so is the most reproducible, but not particularly upgradable. For most, this is the safest way to generate a FiPy environment that actually works.

Comment (wd15) [reply INSTALLATION.rst:72]: Reverse that. Opposite way around.

File: INSTALLATION.rst [hunk 3]
```diff
@@ -61,59 +63,89 @@ Recommended Method
...
+         with any :file:`environments/<SOLVER>-environment.yml` you wish to
+         use (`conda-merge <https://github.com/amitbeka/conda-merge>`_ may
+         prove useful).  Then execute::
 
-     The `fipy conda-forge`_ package is a convenience. You may choose to
-     install packages explicitly, e.g.,::
+           $ conda env create --name <MYFIPYENV> --file <MYMERGEDENVIRONMENT>.yml
 
-       $ conda create --name <MYFIPYENV> --channel conda-forge python=3 numpy scipy matplotlib-base future packaging mpich mpi4py petsc4py mayavi "gmsh <4.0|>=4.5.2"
+    * `conda-lock <https://github.com/conda/conda-lock>`_ lockfiles::
```
Comment (wd15) [inline INSTALLATION.rst:102]: This option will pin all the packages so is the most reproducible, but not particularly upgradable. For most, this is the safest way to generate a FiPy environment that actually works.

File: INSTALLATION.rst (hunk 2 above)
Comment (wd15) [reply INSTALLATION.rst:72]: This option is the most upgradable in the future and probably the best for development.

File: docs/source/ADMINISTRATA.rst (hunk 1 above)
Comment (guyer) [reply docs/source/ADMINISTRATA.rst:129]: It is?

"[A paragraph consisting of two colons ("::") signifies that the following text block(s) comprise a literal block. The literal block must either be indented or quoted (see below).](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#literal-blocks)

**Review by wd15 — APPROVED** [review]

**Comment by wd15** [comment]
Maybe point at a blog post such as [this](https://pythonspeed.com/articles/conda-dependency-management/) to help explain the difference between using lock and not using lock.

**Comment by wd15** [comment]
I see, you've already done it.

**Review by wd15 — APPROVED** [review]

**Comment by wd15** [comment]
BTW the [Binder link](https://www.ctcms.nist.gov/fipy/INSTALLATION.html#pre-installed-on-binder) never seems to work. Might be best to remove that.

**Comment by guyer** [comment]
> BTW the [Binder link](https://www.ctcms.nist.gov/fipy/INSTALLATION.html#pre-installed-on-binder) never seems to work. Might be best to remove that.

https://mybinder.org [seems to be having an outage right now](https://mybinder.readthedocs.io/en/latest/about/status.html), but even when it's up, it might not be too useful. I've added an #987 to take a look at this.

**Pull request merged**